components/script/dom/bindings/htmlconstructor.rs

```diff
@@ -343,3 +352,109 @@ pub fn pop_current_element_queue() {
+        let _ac = JSAutoRealm::new(*cx, new_target.get());
+        if !JS_GetProperty(
+            *cx,
+            new_target.handle(),
+            b"prototype\\0".as_ptr() as *const _,
+            proto_val.handle_mut(),
+        ) {
+            return Err(Error::Type("false".to_owned()));
+        }
+        if !proto_val.is_object() {
+            // Step 7 of https://html.spec.whatwg.org/multipage/#htmlconstructor.
+            // This fallback behavior is designed to match analogous behavior for the
+            // JavaScript built-ins. So we enter the compartment of our underlying
+            // newTarget object and fall back to the prototype object from that global.
+            // XXX The spec says to use GetFunctionRealm(), which is not actually
+            // the same thing as what we have here (e.g. in the case of scripted callable proxies
+            // whose target is not same-compartment with the proxy, or bound functions, etc).
+            // https://bugzilla.mozilla.org/show_bug.cgi?id=1317658
+            rooted!(in(*cx) let global_object = CurrentGlobalOrNull(*cx));
+            get_proto_object(cx, global_object.handle(), prototype.handle_mut());
+        } else {
+            // Step 6
+            prototype.set(proto_val.to_object());
+        };
+    }
+    // Wrap prototype in this context since it is from the newTarget compartment
+    if !JS_WrapObject(*cx, RawMutableHandle::from(prototype.handle_mut())) {
+        return Err(Error::Type("false".to_owned()));
+    }
+    return Ok(());
+}
+
+unsafe fn setup_html_constructor_prototype(
+    cx: JSContext,
+    reflector: HandleObject,
+    prototype: HandleObject,
+) -> bool {
+    rooted!(in(*cx) let mut element = reflector.get());
+    if !JS_WrapObject(*cx, RawMutableHandle::from(element.handle_mut())) {
+        return false;
+    }
+
+    JS_SetPrototype(
+        *cx,
+        RawHandle::from(element.handle()),
+        RawHandle::from(prototype),
+    );
+    return true;
+}
+
+pub unsafe fn call_html_constructor<T: DerivedFrom<Element> + DomObject>(
+    cx: JSContext,
+    args: CallArgs,
+    global: &Window,
+    get_proto_object: fn(JSContext, HandleObject, MutableHandleObject),
+) -> bool {
+    rooted!(in(*cx) let mut prototype = ptr::null_mut::<JSObject>());
+    if let Err(e) = prepare_html_constructor_prototype(cx, args, get_proto_object) {
+        throw_dom_exception(cx, global.upcast(), e);
+        return false;
+    }
+
+    let result: Result<DomRoot<T>, Error> = html_constructor(&global, &args);
+    let result = match result {
+        Ok(result) => result,
+        Err(e) => {
+            throw_dom_exception(cx, global.upcast::<GlobalScope>(), e);
+            return false;
+        },
+    };
+    match setup_html_constructor_prototype(
+        cx,
+        result.reflector().get_jsobject(),
+        prototype.handle(),
+    ) {
+        true => {
+            (result).to_jsval(*cx, MutableHandleValue::from_raw(args.rval()));
```

Comment:
The `(result).to_jsval(...)` is outside of `setup_html_constructor_prototype` because `html_constructor` returns a generic. This prevents having `setup_html_constructor_prototype` being monomorphized as well